### PR TITLE
EEP 74: Erlang Error Index

### PR DESCRIPTION
**Pull request by robertoaloi**
To start the discussion.

### CODE REVIEW

File: eeps/eep-0072.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+Finally, error codes can be used by IDEs (e.g. via language servers)
+to match on error codes and provide contextual help. Both the [Erlang
+LS][] and the [ELP][] language server already use "unofficial" error
+codes.
+
+Format
+------
+
+An error code should be composed by two parts: an alphanumeric
+_namespace_ (three letters) and a numeric identifier (four digits),
```
Comment (robertoaloi) [inline eeps/eep-0072.md]: Should we relax this requirement and allow a dynamic number of chars/digits?

File: eeps/eep-0072.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+The **Erlang Error Index** is a _catalogue_ of errors emitted by
+various tools within the Erlang ecosystem, including - but not limited
+to - the `erlc` Erlang compiler and the `dialyzer` type checker.
+
+The catalogue is not limited to tools shipped with Erlang/OTP, but it
+can include third-party applications such as the [EqWAlizer][]
+type-checker or the [Elvis][] code style reviewer.
+
+Each error in the catalogue is identified by a **unique code** and it
```
Comment (robertoaloi) [inline eeps/eep-0072.md]: Should codes have a human-readable alias? How would we incorporate the "namespace" into the alias?

Comment (jcpetruzza) [reply eeps/eep-0072.md]: I like standardized human-readable aliases; they may not need to be unique (e.g. parsing errors, eqwalizer/dialyzer may report the same thing with a different error)

File: eeps/eep-0072.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+improved over time, error codes remain constant, providing a
+search-engine friendly way to index and reference diagnostics.
+
+An good example of this is the _expression updates a literal_ error
+message, introduced in OTP 27. Given the following code:
+
+    -define(DEFAULT, #{a => 1}).
+
+    updated(Value) ->
+      ?DEFAULT#{a => Value}.
```
Comment (jcpetruzza) [inline eeps/eep-0072.md]: Is `a` meant to be `timeout`?

File: eeps/eep-0072.md [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+Required Changes
+================
+
+* The `erlc` compiler and the `dialyzer` type-checker would start
+  emitting error codes together with errors and warnings messags.
+* The tools could also expose a `--explain` option from the command
+  line - similarly to `rustc` - which would take an error code as the
+  input and return the Markdown documentation as output.
+
```
Comment (jcpetruzza) [inline eeps/eep-0072.md]: Would it make sense to extend the proposal to include a machine-readable format (JSON?) for reporting errors

File: eeps/eep-0072.md [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+  list of _official_ namespaces. Each tool maintainer would then be
+  responsible to allocate specific codes to specific diagnostics.
+
+Processes
+---------
+
+Error cannot be re-used. If a tool stops emitting an error code, the
+_deprecated_ error code is still documented in the index, together
+with a deprecation notice. This is to avoid re-using a single code for
+multiple purposes.
```
Comment (jcpetruzza) [inline eeps/eep-0072.md]: Maybe propose a mechanism to submit new namespaces/codes?

File: eeps/eep-0072.md [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+The website should use _Markdown_ as the primary mechanism to write
+content and it should be easily extensible by the Community.
+
+Required Changes
+================
+
+* The `erlc` compiler and the `dialyzer` type-checker would start
+  emitting error codes together with errors and warnings messags.
+* The tools could also expose a `--explain` option from the command
```
Comment (garazdawi) [inline eeps/eep-0072.md]: How would `erlc` fetch the full explanation? How does `rustc` do it?

File: eeps/eep-0072.md (hunk 1 above)
Comment (garazdawi) [reply eeps/eep-0072.md]: We can relax it later if needed.

File: eeps/eep-0072.md (hunk 6 above)
Comment (robertoaloi) [reply eeps/eep-0072.md]: @garazdawi I had to check. Apparently the Markdown files are actually stored in a crate which is part of the compiler itself:

https://github.com/rust-lang/rust/blob/master/compiler/rustc_error_codes/src/error_codes/

Then they have a [registry](https://github.com/rust-lang/rust/blob/master/compiler/rustc_errors/src/registry.rs) (a hashmap) which maps each error code to its corresponding description.

When you [pass](https://github.com/rust-lang/rust/blob/edbd9398dcc0e3b6ffe0dd0cd4578d37d1d36974/compiler/rustc_driver_impl/src/lib.rs#L307) the `explain` command they simply [try to fetch](https://github.com/rust-lang/rust/blob/edbd9398dcc0e3b6ffe0dd0cd4578d37d1d36974/compiler/rustc_driver_impl/src/lib.rs#L536) the description. Not very advanced.

The initial plumbing was introduced [here](https://github.com/rust-lang/rust/commit/9b9cce2316119a2ffdc9556d410e464b7542d64d).

File: eeps/eep-0072.md (hunk 4 above)
Comment (garazdawi) [reply eeps/eep-0072.md]: IMO that is a separate issue. This EEP should focus on the index and how to allocate names/codes to it.

File: eeps/eep-0072.md (hunk 6 above)
Comment (garazdawi) [reply eeps/eep-0072.md]: IMO we should remove this as it creates a dependency between what is stored online and what is embedded in the compiler. It is better if everyone just uses the webpage directly.

**Review by garazdawi — COMMENTED** [review]
I looked at how haskell is doing things and I like it. Simple and effective.

What I think we should do is add https://errors.erlang.org as a github page. That github page will contain errors for the tools part of Erlang/OTP, but it will also contain links to other namespaces so that users can find the errors for elvis, elp, elixir, gradualizer etc etc from that page. I don't think we should make the external errors part of that page as that would mean that the amount of administration needed will increase. Possibly we can make the search index incorporate all namespaces.

I then propose that it will be a joint effort from us and WhatsApp to manage this page, that is we will all have merge/commit rights in that repo.

Maybe we should also create a convention on how errors/warnings are printed so that tools know how to parse them?

File: eeps/eep-0072.md (hunk 3 above)
Comment (robertoaloi) [reply eeps/eep-0072.md]: Yes, fixed.

**Comment by robertoaloi** [comment]
Superseded by https://github.com/erlang/eep/pull/70

**Pull request closed without merging**